src/models/prompter.js

```diff
@@ -295,29 +304,102 @@ export class Prompter {
+    //         prompt = await this.replaceStrings(prompt, messages, this.convo_examples);
+    //         let generation = await this.chat_model.sendRequest(messages, prompt);
+    //         // in conversations >2 players LLMs tend to hallucinate and role-play as other bots
+    //         // the FROM OTHER BOT tag should never be generated by the LLM
+    //         if (generation.includes('(FROM OTHER BOT)')) {
+    //             console.warn('LLM hallucinated message as another bot. Trying again...');
+    //             continue;
+    //         }
+    //         if (current_msg_time !== this.most_recent_msg_time) {
+    //             console.warn(this.agent.name + ' received new message while generating, discarding old response.');
+    //             return '';
+    //         }
+    //         return generation;
+    //     }
+    //     return '';
+    // }
+
+    async saveToFile(logFile, logEntry) {
+        let task_id = this.agent.task.task_id;
+        console.log(task_id)
+        let logDir;
+        if (this.task_id === null) {
+            logDir = path.join(__dirname, `../../bots/${this.agent.name}/logs`);
+        } else {
+            logDir = path.join(__dirname, `../../bots/${this.agent.name}/logs/${task_id}`);
+        }
+
+        await fs.mkdir(logDir, { recursive: true });
+
+        logFile = path.join(logDir, logFile);
+        await fs.appendFile(logFile, String(logEntry), 'utf-8');
+    }
+
     async promptConvo(messages) {
+        // console.log(`[${new Date().toISOString()}] promptConvo called with messages:`, messages);
+
         this.most_recent_msg_time = Date.now();
         let current_msg_time = this.most_recent_msg_time;
+
         for (let i = 0; i < 3; i++) { // try 3 times to avoid hallucinations
             await this.checkCooldown();
             if (current_msg_time !== this.most_recent_msg_time) {
                 return '';
             }
+
             let prompt = this.profile.conversing;
             prompt = await this.replaceStrings(prompt, messages, this.convo_examples);
-            let generation = await this.chat_model.sendRequest(messages, prompt);
-            // in conversations >2 players LLMs tend to hallucinate and role-play as other bots
-            // the FROM OTHER BOT tag should never be generated by the LLM
-            if (generation.includes('(FROM OTHER BOT)')) {
+            let generation;
+
+            try {
+                generation = await this.chat_model.sendRequest(messages, prompt);
+                let task_id = this.agent.task.task_id
+                console.log(task_id)
+                if (typeof generation !== 'string') {
+                    console.error('Error: Generated response is not a string', generation);
+                    throw new Error('Generated response is not a string');
+                }
+                console.log("Generated response:", generation); 
+                const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
```

Comment:
move this to its own function so its cleaner and we can reuse in other prompt funcs